plugins/incremental-ingestion-backend/src/database/IncrementalIngestionDatabaseManager.ts

```diff
@@ -0,0 +1,592 @@
+/*
+ * Copyright 2022 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { Knex } from 'knex';
+import type { DeferredEntity } from '@backstage/plugin-catalog-backend';
+import { stringifyEntityRef } from '@backstage/catalog-model';
+import { Duration } from 'luxon';
+import { v4 } from 'uuid';
+import {
+  INCREMENTAL_ENTITY_PROVIDER_ANNOTATION,
+  IngestionRecord,
+  IngestionRecordUpdate,
+  IngestionUpsertIFace,
+  MarkRecord,
+  MarkRecordInsert,
+} from '../types';
+
+/** @public */
+export class IncrementalIngestionDatabaseManager {
+  private client: Knex;
+
+  constructor(options: { client: Knex }) {
+    this.client = options.client;
+  }
+
+  /**
+   * Performs an update to the ingestion record with matching `id`.
+   * @param options - IngestionRecordUpdate
+   */
+  async updateIngestionRecordById(options: IngestionRecordUpdate) {
+    await this.client.transaction(async tx => {
+      const { ingestionId, update } = options;
+      await tx('ingestion.ingestions').where('id', ingestionId).update(update);
+    });
+  }
+
+  /**
+   * Performs an update to the ingestion record with matching provider name. Will only update active records.
+   * @param provider - string
+   * @param update - Partial<IngestionUpsertIFace>
+   */
+  async updateIngestionRecordByProvider(
+    provider: string,
+    update: Partial<IngestionUpsertIFace>,
+  ) {
+    await this.client.transaction(async tx => {
+      await tx('ingestion.ingestions')
+        .where('provider_name', provider)
+        .andWhere('completion_ticket', 'open')
+        .update(update);
+    });
+  }
+
+  /**
+   * Performs an insert into the `ingestion.ingestions` table with the supplied values.
+   * @param record - IngestionUpsertIFace
+   */
+  async insertIngestionRecord(record: IngestionUpsertIFace) {
+    await this.client.transaction(async tx => {
+      await tx('ingestion.ingestions').insert(record);
+    });
+  }
+
+  private async deleteMarkEntities(
+    tx: Knex.Transaction,
+    ids: { id: string }[],
+  ) {
+    const chunks: { id: string }[][] = [];
+    for (let i = 0; i < ids.length; i += 100) {
+      const chunk = ids.slice(i, i + 100);
```

Comment:
there's also `lodash.chunk` if you want

Reply:
I think maybe at some point, but for now it seems like a micro-optimization.

Reply:
yeah not an optimization at all, just not having to write (and maybe get some bug) the code by hand